From 266c691c8019aa5e815644e218d1a12e01f91eb6 Mon Sep 17 00:00:00 2001
From: Flopster101 <nahuelgomez329@gmail.com>
Date: Wed, 21 Jan 2026 15:46:46 +0300
Subject: [PATCH 2/2] Settings: Add SBC HD Developer Options toggles

Add two new toggles in Developer Options for SBC HD audio:
- Enable SBC HD audio: Enables Dual Channel mode with higher bitrates
- Force max SBC HD bitrate: Forces 551 kbps even without 3Mbps EDR

Both settings require toggling Bluetooth to take effect.

Change-Id: I339d6055c1cb78f38e90561ea8e77ea528ce095e
---
 res/xml/development_settings.xml              | 10 +++
 ...uetoothSbcHdForceMaxBitrateController.java | 71 +++++++++++++++++++
 .../BluetoothSbcHdPreferenceController.java   | 71 +++++++++++++++++++
 .../DevelopmentSettingsDashboardFragment.java |  2 +
 4 files changed, 154 insertions(+)
 create mode 100644 src/com/android/settings/development/BluetoothSbcHdForceMaxBitrateController.java
 create mode 100644 src/com/android/settings/development/BluetoothSbcHdPreferenceController.java

diff --git a/res/xml/development_settings.xml b/res/xml/development_settings.xml
index aba7a70..859a3b1 100644
--- a/res/xml/development_settings.xml
+++ b/res/xml/development_settings.xml
@@ -395,6 +395,16 @@
             android:title="@string/bluetooth_disable_absolute_volume"
             android:summary="@string/bluetooth_disable_absolute_volume_summary" />
 
+        <SwitchPreferenceCompat
+            android:key="bluetooth_enable_sbc_hd"
+            android:title="@string/bluetooth_enable_sbc_hd"
+            android:summary="@string/bluetooth_enable_sbc_hd_summary" />
+
+        <SwitchPreferenceCompat
+            android:key="bluetooth_sbc_hd_force_max_bitrate"
+            android:title="@string/bluetooth_sbc_hd_force_max_bitrate"
+            android:summary="@string/bluetooth_sbc_hd_force_max_bitrate_summary" />
+
         <SwitchPreferenceCompat
             android:key="bluetooth_disable_leaudio"
             android:title="@string/bluetooth_disable_leaudio"
diff --git a/src/com/android/settings/development/BluetoothSbcHdForceMaxBitrateController.java b/src/com/android/settings/development/BluetoothSbcHdForceMaxBitrateController.java
new file mode 100644
index 0000000..63ae82f
--- /dev/null
+++ b/src/com/android/settings/development/BluetoothSbcHdForceMaxBitrateController.java
@@ -0,0 +1,71 @@
+/*
+ * Copyright (C) 2024 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.settings.development;
+
+import android.content.Context;
+import android.os.SystemProperties;
+
+import androidx.annotation.VisibleForTesting;
+import androidx.preference.Preference;
+import androidx.preference.TwoStatePreference;
+
+import com.android.settings.core.PreferenceControllerMixin;
+import com.android.settingslib.development.DeveloperOptionsPreferenceController;
+
+/**
+ * Controller for forcing maximum SBC HD bitrate (551 kbps at 44.1kHz).
+ * This bypasses the 3Mbps EDR check and forces 3DH5 packets even if the
+ * device doesn't advertise support. May cause audio dropouts.
+ */
+public class BluetoothSbcHdForceMaxBitrateController extends
+        DeveloperOptionsPreferenceController implements Preference.OnPreferenceChangeListener,
+        PreferenceControllerMixin {
+
+    private static final String BLUETOOTH_SBC_HD_FORCE_MAX_KEY = "bluetooth_sbc_hd_force_max_bitrate";
+    @VisibleForTesting
+    static final String BLUETOOTH_SBC_HD_FORCE_MAX_PROPERTY = "persist.bluetooth.sbc_hd.force_max_bitrate";
+
+    public BluetoothSbcHdForceMaxBitrateController(Context context) {
+        super(context);
+    }
+
+    @Override
+    public String getPreferenceKey() {
+        return BLUETOOTH_SBC_HD_FORCE_MAX_KEY;
+    }
+
+    @Override
+    public boolean onPreferenceChange(Preference preference, Object newValue) {
+        final boolean isEnabled = (Boolean) newValue;
+        SystemProperties.set(BLUETOOTH_SBC_HD_FORCE_MAX_PROPERTY, isEnabled ? "true" : "false");
+        return true;
+    }
+
+    @Override
+    public void updateState(Preference preference) {
+        final boolean isEnabled = SystemProperties.getBoolean(
+                BLUETOOTH_SBC_HD_FORCE_MAX_PROPERTY, false /* default */);
+        ((TwoStatePreference) mPreference).setChecked(isEnabled);
+    }
+
+    @Override
+    protected void onDeveloperOptionsSwitchDisabled() {
+        super.onDeveloperOptionsSwitchDisabled();
+        SystemProperties.set(BLUETOOTH_SBC_HD_FORCE_MAX_PROPERTY, "false");
+        ((TwoStatePreference) mPreference).setChecked(false);
+    }
+}
diff --git a/src/com/android/settings/development/BluetoothSbcHdPreferenceController.java b/src/com/android/settings/development/BluetoothSbcHdPreferenceController.java
new file mode 100644
index 0000000..a6b5dd7
--- /dev/null
+++ b/src/com/android/settings/development/BluetoothSbcHdPreferenceController.java
@@ -0,0 +1,71 @@
+/*
+ * Copyright (C) 2024 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.settings.development;
+
+import android.content.Context;
+import android.os.SystemProperties;
+
+import androidx.annotation.VisibleForTesting;
+import androidx.preference.Preference;
+import androidx.preference.TwoStatePreference;
+
+import com.android.settings.core.PreferenceControllerMixin;
+import com.android.settingslib.development.DeveloperOptionsPreferenceController;
+
+/**
+ * Controller for enabling SBC HD (Dual Channel) Bluetooth audio mode.
+ * When enabled, SBC will automatically use Dual Channel mode with higher bitrates,
+ * prioritized over AAC. Requires toggling Bluetooth to take effect.
+ */
+public class BluetoothSbcHdPreferenceController extends
+        DeveloperOptionsPreferenceController implements Preference.OnPreferenceChangeListener,
+        PreferenceControllerMixin {
+
+    private static final String BLUETOOTH_SBC_HD_KEY = "bluetooth_enable_sbc_hd";
+    @VisibleForTesting
+    static final String BLUETOOTH_SBC_HD_PROPERTY = "persist.bluetooth.sbc_hd.enabled";
+
+    public BluetoothSbcHdPreferenceController(Context context) {
+        super(context);
+    }
+
+    @Override
+    public String getPreferenceKey() {
+        return BLUETOOTH_SBC_HD_KEY;
+    }
+
+    @Override
+    public boolean onPreferenceChange(Preference preference, Object newValue) {
+        final boolean isEnabled = (Boolean) newValue;
+        SystemProperties.set(BLUETOOTH_SBC_HD_PROPERTY, isEnabled ? "true" : "false");
+        return true;
+    }
+
+    @Override
+    public void updateState(Preference preference) {
+        final boolean isEnabled = SystemProperties.getBoolean(
+                BLUETOOTH_SBC_HD_PROPERTY, false /* default */);
+        ((TwoStatePreference) mPreference).setChecked(isEnabled);
+    }
+
+    @Override
+    protected void onDeveloperOptionsSwitchDisabled() {
+        super.onDeveloperOptionsSwitchDisabled();
+        SystemProperties.set(BLUETOOTH_SBC_HD_PROPERTY, "false");
+        ((TwoStatePreference) mPreference).setChecked(false);
+    }
+}
diff --git a/src/com/android/settings/development/DevelopmentSettingsDashboardFragment.java b/src/com/android/settings/development/DevelopmentSettingsDashboardFragment.java
index 443d6c2..a2626d8 100644
--- a/src/com/android/settings/development/DevelopmentSettingsDashboardFragment.java
+++ b/src/com/android/settings/development/DevelopmentSettingsDashboardFragment.java
@@ -762,6 +762,8 @@ public class DevelopmentSettingsDashboardFragment extends RestrictedDashboardFra
         controllers.add(new TetheringHardwareAccelPreferenceController(context));
         controllers.add(new BluetoothDeviceNoNamePreferenceController(context));
         controllers.add(new BluetoothAbsoluteVolumePreferenceController(context));
+        controllers.add(new BluetoothSbcHdPreferenceController(context));
+        controllers.add(new BluetoothSbcHdForceMaxBitrateController(context));
         controllers.add(new BluetoothAvrcpVersionPreferenceController(context));
         controllers.add(new BluetoothMapVersionPreferenceController(context));
         controllers.add(new BluetoothLeAudioPreferenceController(context, fragment));
-- 
2.54.0

